### PR TITLE
Add warning if client/server version difference exceeds the supported skew

### PR DESCRIPTION
**Pull request by brianpursley**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR writes a warning message to stderr when you run `kubectl version` if the version skew between the client and server is more than the supported skew (+/- 1 minor version).

Example warning message:
```
WARNING: version difference between client (1.21+) and server (1.19) exceeds the supported minor version skew of +/-1
```

**Which issue(s) this PR fixes**:

https://github.com/kubernetes/kubectl/issues/685

**Special notes for your reviewer**:
The approach of only warning when the user runs `kubectl version` was discussed at the 1/13/2021 SIG-CLI meeting.

Long-term we may want to introduce a more intrusive warning to alert users of an unsupported skew, but that is beyond the scope of this PR.

**Does this PR introduce a user-facing change?**:
```release-note
kubectl version changed to write a warning message to stderr if the client and server version difference exceeds the supported version skew of +/-1 minor version.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go:2]: nit: we're in 2021 :D

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go:27]: Do we have a consensus into this? To be honest, as Kubernetes supports at least 3 minor versions, (like from 1.20 to 1.18) I'm afraid of maybe this issuing a warning because of a small difference, but this is my opinion :)

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go:52]: To be honest (and this probably is because I'm pretty a serial reader...) this seemed confused to me :)

It might be worst but more "readable", maybe having a math.Abs(int(clientVersion - serverVersion)) and checking if this is bigger then the Skew :)

But I might be wrong, or maybe just bikeshedding here so feel free to ignore :)

**Review by rikatz — COMMENTED** [review]
Left some comments just as a second opinion :)

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go
Comment (pwittrock) [reply staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go:27]: +-1 version does support 3 versions -- 1.19 supports 1.18, 1.19, 1.20

Comment (rikatz) [reply staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go:27]: True! Will re-read what I've wrote on the other review to see if still makes sense

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go:2]: Whoops, I guess 2020 was so good I forgot to change it :)

Thanks for catching that

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go:31]: Let me know if you determine something different.  I added a comment explaining which versions would fall within the supported skew range.

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go
Comment (dougsland) [inline staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go:58]: I am wondering if for re-use in different places (when required) parseMajorMinorVersion() should belong to version.go.

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go
Comment (dougsland) [inline staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go:42]: The functions are pretty clear but having doc string always help.

**Review by dougsland — COMMENTED** [review]
I have few comments. However, I must say: LGTM.

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go:39]: Added doc strig

**Comment by brianpursley** [comment]
Thanks to a comment @soltysh left in another PR, I became aware of the existence of [ParseSemantic](https://github.com/kubernetes/kubernetes/blob/f170d33e1f320d7a3bcd2ae5a60b7d335ae382f6/staging/src/k8s.io/apimachinery/pkg/util/version/version.go#L111) in api machinery.  I updated this PR to use this to get the major and minor versions instead of reinventing the wheel, so to speak.

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go]: Thanks for this comment. I found an existing function in API machinery to get the versions for comparison instead of writing my own.
https://github.com/kubernetes/kubernetes/blob/f170d33e1f320d7a3bcd2ae5a60b7d335ae382f6/staging/src/k8s.io/apimachinery/pkg/util/version/version.go#L111

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go]: Yeah, I don’t really know what else to say here. It’s really not horrible to use an out of skew client, but if you are running into problems, you should be aware that it could be an unsupported skew.

A main reason for adding this warning is to make people aware of the unsupported skew before they open an issue in the kubectl repo.

but you’re right, it is not super actionable.

**Comment by brianpursley** [comment]
@dougsland See my comment above, I was able to re-use an existing function to get major/minor version, so the code has changed.  Can you take another look?  Thanks!

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go
Comment (dougsland) [reply staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go]: Hey @briandorsey I believe the warning you created is super valid but it's difficult to avoid users to open issues with unsupported skew IMO. 

Probably we might have such scenarios with users out there:
  - didn't understand the message - what's skew +/- 1? 
  - simple ignore it as they don't know what to do next 
  - not sure "how to fix it"
  - not sure "how to make the cluster in a minimum good shape" to open the issue. 
   
From my side, the code looks great but I would request more people to review the message to have more ideas. Probably a good PR for request ideas in the SIG-CLI meeting.

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning_test.go
Comment (seans3) [inline staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning_test.go:43]: Nit: is the version supposed to be "v1.19.7-alpha" ?

**Review by seans3 — APPROVED** [review]
This looks good to me. I actually think the warning message is about as good as it can be. I've added the v1.21 milestone, since we hope to get this in by code freeze. I can LGTM once Maciej is happy (and the BUILD file is removed).

File: staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/cmd/version/skew_warning.go:24]: Nit: but try to keep imports split into:
1. built-in 
2. 3rd party
3. k8s

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
